backend/services/event.py

```diff
@@ -189,3 +233,296 @@ def delete(self, subject: User, id: int) -> None:
+        """
+        Register a user for an event.
+
+        Args:
+            subject: User making the registration request
+            attendee: The user being registered for the event
+            event: The EventDetails being registered for
+
+        Returns:
+            EventRegistration
+
+        Raises:
+            UserPermissionException if subject does not have permission to register user
+            EventRegistrationException if the event is full
+        """
+
+        # Get the registration status.
+        # NOTE: It is preferred to use the service function rather than the list of
+        # registrations passed in from `event` in the case that registrations are added
+        # between when `event` was fetched and this function runs.
+        event_status = self.get_event_registration_status(event.id)
+
+        # Raise exception if event is full.
+        if event_status.registration_count >= event.registration_limit:
+            raise EventRegistrationException(event.id)
+
+        # Enable idemopotency in returning existing registration, if one exists.
+        # Permission to manage / read registration is enforced in EventService#get_registration
+        existing_registration = self.get_registration(subject, attendee, event)
+        if existing_registration:
+            return existing_registration
+
+        # Add new object to table and commit changes
+        event_registration_entity = EventRegistrationEntity(
+            user_id=attendee.id, event_id=event.id
+        )
+        self._session.add(event_registration_entity)
+        self._session.commit()
+
+        # Return registration
+        return event_registration_entity.to_model()
+
+    def unregister(self, subject: User, attendee: User, event: EventDetails) -> None:
+        """
+        Delete a user's event registration.
+
+        Args:
+            subject: User performing the unregister action
+            attendee: User whose registration is being deleted
+            event: the event the attendee is unregistering for
+
+        Returns:
+            None in a successful invocation. Idempotent in the case of not registered.
+
+        Raises:
+            UserPermissionException when the user is not authorized to manage the registration.
+        """
+        # Find registration to delete
+        # Permissions for reading/managing registration are enfoced in #get_registration
+        event_registration = self.get_registration(subject, attendee, event)
+
+        # Ensure object exists
+        if event_registration is None:
+            raise ResourceNotFoundException(
+                f"No event registration found with matching event ID: {id}"
+            )
+
+        # Delete object and commit
+        self._session.delete(
+            self._session.get(
+                EventRegistrationEntity,
+                (event_registration.event.id, event_registration.user.id),
+            )
+        )
+        self._session.commit()

```

Comment:
To make this idempotent, as advertised in the docstring, how about if `event_registration` is None we just return early rather than raise?